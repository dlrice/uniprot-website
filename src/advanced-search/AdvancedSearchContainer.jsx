import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  selectField,
  updateInputValue,
  updateEvidence,
  updateRangeValue,
  updateLogicOperator,
  addClause,
  removeClause,
  fetchSearchTerms,
  fetchEvidencesIfNeeded,
} from './state/actions';
import { copyQueryObjectGoToResults } from '../state/actions';
import ClauseList from './ClauseList';

export class AdvancedSearch extends Component {
  componentDidMount() {
    const { dispatchFetchSearchTerms, dispatchfetchEvidencesIfNeeded } = this.props;
    dispatchFetchSearchTerms();
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
  }

  render() {
    const { namespace, dispatchAddClause, dispatchCopyQueryObjectGoToResults } = this.props;
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        <ClauseList {...this.props} />
        <hr />
        <div>
          <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
            Add Field
          </button>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={dispatchCopyQueryObjectGoToResults}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
});

const mapDispatchToProps = dispatch => ({
  handleFieldSelect: (clauseId, field) => dispatch(selectField(clauseId, field)),
  handleInputChange: (clauseId, value) => dispatch(updateInputValue(clauseId, value)),
  handleEvidenceChange: (clauseId, value) => dispatch(updateEvidence(clauseId, value)),
  handleRangeInputChange: (clauseId, value, from) => dispatch(updateRangeValue(clauseId, value, from)),
  handleLogicChange: (clauseId, value) => dispatch(updateLogicOperator(clauseId, value)),
  handleRemoveClause: clauseId => dispatch(removeClause(clauseId)),
  dispatchAddClause: () => dispatch(addClause()),
  dispatchfetchEvidencesIfNeeded: evidencesType => dispatch(fetchEvidencesIfNeeded(evidencesType)),
  dispatchFetchSearchTerms: () => dispatch(fetchSearchTerms()),
  dispatchCopyQueryObjectGoToResults: () => dispatch(copyQueryObjectGoToResults()),
});

const AdvancedSearchContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearch);

export default AdvancedSearchContainer;
