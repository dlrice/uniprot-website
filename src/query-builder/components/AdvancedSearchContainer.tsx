import React, { Component, FormEvent, MouseEvent } from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { PageIntro } from 'franklin-sites';

import {
  Clause,
  SearchTermType,
  Operator,
  Namespace,
} from '../../uniprotkb/types/searchTypes';

import AdvancedSearch from './AdvancedSearch';

import { RootState, RootAction } from '../../app/state/rootInitialState';
import * as searchActions from '../../uniprotkb/state/searchActions';

import { stringify } from '../utils/queryStringProcessor';

import { Location, LocationToPath } from '../../app/config/urls';

import '../../uniprotkb/components/search/styles/search-container.scss';

const queryBuilderPath = LocationToPath[Location.UniProtKBQueryBuilder];

type Props = {
  dispatchUpdateQueryString: (type: string) => void;
  namespace: Namespace;
  clauses: Clause[];
  dispatchUpdateClauses: (clauses: Clause[]) => void;
  dispatchAddClause: () => void;
  handleFieldSelect: (clauseId: string, field: SearchTermType) => void;
  handleInputChange: (clauseId: string, value: string, id?: string) => void;
  handleEvidenceChange: (clauseId: string, value: string) => void;
  handleRangeInputChange: (
    clauseId: string,
    value: string,
    from?: boolean
  ) => void;
  handleLogicChange: (clauseId: string, value: Operator) => void;
  handleRemoveClause: (clauseId: string) => void;
  dispatchSetPreSelectedClauses: () => void;
} & RouteComponentProps;

type State = {
  queryString: string;
};

export class Search extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    // const { queryString } = props;
    // this.state = { queryString };
    this.state = { queryString: '' };
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleQueryStringChange = this.handleQueryStringChange.bind(this);
  }

  componentDidMount() {
    const {
      dispatchSetPreSelectedClauses,
      location: { pathname },
      history,
    } = this.props;
    if (queryBuilderPath && pathname === `${queryBuilderPath}/reset`) {
      dispatchSetPreSelectedClauses();
      history.replace(queryBuilderPath);
    }
  }

  handleSubmitClick(event: FormEvent | MouseEvent) {
    event.preventDefault();

    const { history, clauses, dispatchUpdateQueryString } = this.props;
    const queryString = stringify(clauses);
    dispatchUpdateQueryString(queryString);
    history.push({
      pathname: '/uniprotkb',
      search: queryString && `query=${queryString}`,
    });
  }

  handleQueryStringChange(queryString: string) {
    this.setState({ queryString });
  }

  render() {
    const { queryString } = this.state;

    return (
      <>
        <PageIntro title="Advanced search" />
        <AdvancedSearch
          {...this.props}
          queryString={queryString}
          handleAdvancedSubmitClick={this.handleSubmitClick}
        />
      </>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  clauses: state.query.clauses,
  namespace: state.query.namespace,
});

const mapDispatchToProps = (dispatch: Dispatch<RootAction>) =>
  bindActionCreators(
    {
      handleFieldSelect: (clauseId: string, field: SearchTermType) =>
        searchActions.selectSearchTerm(clauseId, field),
      handleInputChange: (clauseId: string, value: string, id?: string) =>
        searchActions.updateInputValue(clauseId, value, id),
      handleEvidenceChange: (clauseId: string, value: string) =>
        searchActions.updateEvidence(clauseId, value),
      handleRangeInputChange: (
        clauseId: string,
        value: string,
        from?: boolean
      ) => searchActions.updateRangeValue(clauseId, value, from),
      handleLogicChange: (clauseId: string, value: Operator) =>
        searchActions.updateLogicOperator(clauseId, value),
      handleRemoveClause: (clauseId: string) =>
        searchActions.removeClause(clauseId),
      dispatchAddClause: () => searchActions.addClause(),
      dispatchSubmitAdvancedQuery: () => searchActions.submitAdvancedQuery(),
      dispatchUpdateClauses: (clauses) => searchActions.updateClauses(clauses),
      dispatchUpdateQueryString: (queryString) =>
        searchActions.updateQueryString(queryString),
      dispatchSetPreSelectedClauses: () =>
        searchActions.setPreSelectedClauses(),
    },
    dispatch
  );

const AdvancedSearchContainer = withRouter(
  connect(mapStateToProps, mapDispatchToProps)(Search)
);

export default AdvancedSearchContainer;
