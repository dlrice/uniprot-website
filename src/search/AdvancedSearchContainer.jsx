import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { MainSearch } from 'franklin-sites';
import {
  selectField,
  updateInputValue,
  updateEvidence,
  updateRangeValue,
  updateLogicOperator,
  updateQueryString,
  submitAdvancedQuery,
  addClause,
  removeClause,
  fetchSearchTerms,
  fetchEvidencesIfNeeded,
} from './state/actions';
import ClauseList from './ClauseList';

export class AdvancedSearch extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showAdvanced: false,
    };
    this.toggleAdvanced = this.toggleAdvanced.bind(this);
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleAdvancedSubmitClick = this.handleAdvancedSubmitClick.bind(this);
  }

  componentDidMount() {
    const { dispatchFetchSearchTerms, dispatchfetchEvidencesIfNeeded } = this.props;
    dispatchFetchSearchTerms();
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
  }

  toggleAdvanced() {
    const { showAdvanced } = this.state;
    this.setState({ showAdvanced: !showAdvanced });
  }

  handleAdvancedSubmitClick() {
    const { dispatchSubmitAdvancedQuery, history } = this.props;
    dispatchSubmitAdvancedQuery();
    history.push('/uniprotkb');
  }

  handleSubmitClick() {
    const { dispatchQueryStringChange, history } = this.props;
    dispatchQueryStringChange();
    history.push('/uniprotkb');
  }

  render() {
    const { namespace, dispatchAddClause, queryString } = this.props;
    const { showAdvanced } = this.state;
    let search;
    if (showAdvanced) {
      search = (
        <Fragment>
          <ClauseList {...this.props} />
          <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
            Add Field
          </button>
        </Fragment>
      );
    } else {
      search = <MainSearch handleSearchSubmit={this.handleSubmitClick} searchTerm={queryString} />;
    }

    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        {search}
        <button type="button" onClick={this.toggleAdvanced}>
          Advanced search
        </button>
        <div>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={this.handleAdvancedSubmitClick}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  clauses: state.query.clauses,
  queryString: state.query.queryString,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
});

const mapDispatchToProps = dispatch => ({
  handleFieldSelect: (clauseId, field) => dispatch(selectField(clauseId, field)),
  handleInputChange: (clauseId, value) => dispatch(updateInputValue(clauseId, value)),
  handleEvidenceChange: (clauseId, value) => dispatch(updateEvidence(clauseId, value)),
  handleRangeInputChange: (clauseId, value, from) => dispatch(updateRangeValue(clauseId, value, from)),
  handleLogicChange: (clauseId, value) => dispatch(updateLogicOperator(clauseId, value)),
  handleRemoveClause: clauseId => dispatch(removeClause(clauseId)),
  dispatchAddClause: () => dispatch(addClause()),
  dispatchfetchEvidencesIfNeeded: evidencesType => dispatch(fetchEvidencesIfNeeded(evidencesType)),
  dispatchFetchSearchTerms: () => dispatch(fetchSearchTerms()),
  dispatchSubmitAdvancedQuery: () => dispatch(submitAdvancedQuery()),
  dispatchQueryStringChange: queryString => dispatch(updateQueryString(queryString)),
});

const AdvancedSearchContainer = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(AdvancedSearch),
);

export default AdvancedSearchContainer;
