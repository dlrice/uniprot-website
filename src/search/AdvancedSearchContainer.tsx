import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { RootState, RootAction } from '../state/state-types.d';
import * as searchActions from './state/searchActions';
import {
  Clause,
  SearchTermType,
  Operator,
  Evidence,
  Evidences,
  Namespace,
} from './types/searchTypes';
import AdvancedSearch from './AdvancedSearch';
import { createQueryString } from './utils/QueryStringGenerator';

import './styles/SearchContainer.scss';

type Props = {
  queryString: string;
  dispatchUpdateQueryString: (type: string) => void;
  searchTerms: SearchTermType[];
  namespace: Namespace;
  clauses: Clause[];
  evidences: Evidences;
  dispatchUpdateClauses: (clauses: Clause[]) => void;
  dispatchfetchEvidencesIfNeeded: (type: Evidence) => void;
  dispatchFetchSearchTermsIfNeeded: () => void;
  dispatchAddClause: () => void;
  handleFieldSelect: (clauseId: string, field: SearchTermType) => void;
  handleInputChange: (clauseId: string, value: string, id?: string) => void;
  handleEvidenceChange: (clauseId: string, value: string) => void;
  handleRangeInputChange: (
    clauseId: string,
    value: string,
    from?: boolean
  ) => void;
  handleLogicChange: (clauseId: string, value: Operator) => void;
  handleRemoveClause: (clauseId: string) => void;
} & RouteComponentProps;

type State = {
  queryString: string;
};

export class Search extends Component<Props, State> {
  public constructor(props: Props) {
    super(props);
    const { queryString } = props;
    this.state = { queryString };
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleQueryStringChange = this.handleQueryStringChange.bind(this);
  }

  public componentDidUpdate(prevProps: Props) {
    const { queryString: prevQueryString } = prevProps;
    const { queryString } = this.props;
    if (prevQueryString !== queryString) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ queryString });
    }
  }

  private handleSubmitClick() {
    const { history, clauses, dispatchUpdateQueryString } = this.props;
    const queryString = createQueryString(clauses);
    dispatchUpdateQueryString(queryString);
    history.push({ pathname: '/uniprotkb', search: `query=${queryString}` });
  }

  private handleQueryStringChange(queryString: string) {
    this.setState({ queryString });
  }

  public render() {
    const { queryString } = this.state;
    const search = (
      <AdvancedSearch
        {...this.props}
        queryString={queryString}
        handleAdvancedSubmitClick={this.handleSubmitClick}
      />
    );
    return <Fragment>{search}</Fragment>;
  }
}

const mapStateToProps = (state: RootState) => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
  queryString: state.query.queryString,
});

const mapDispatchToProps = (dispatch: Dispatch<RootAction>) =>
  bindActionCreators(
    {
      handleFieldSelect: (clauseId: string, field: SearchTermType) =>
        searchActions.selectSearchTerm(clauseId, field),
      handleInputChange: (clauseId: string, value: string, id?: string) =>
        searchActions.updateInputValue(clauseId, value, id),
      handleEvidenceChange: (clauseId: string, value: string) =>
        searchActions.updateEvidence(clauseId, value),
      handleRangeInputChange: (
        clauseId: string,
        value: string,
        from?: boolean
      ) => searchActions.updateRangeValue(clauseId, value, from),
      handleLogicChange: (clauseId: string, value: Operator) =>
        searchActions.updateLogicOperator(clauseId, value),
      handleRemoveClause: (clauseId: string) =>
        searchActions.removeClause(clauseId),
      dispatchAddClause: () => searchActions.addClause(),
      dispatchfetchEvidencesIfNeeded: evidencesType =>
        searchActions.fetchEvidencesIfNeeded(evidencesType),
      dispatchFetchSearchTermsIfNeeded: () =>
        searchActions.fetchSearchTermsIfNeeded(),
      dispatchSubmitAdvancedQuery: () => searchActions.submitAdvancedQuery(),
      dispatchUpdateClauses: clauses => searchActions.updateClauses(clauses),
      dispatchUpdateQueryString: queryString =>
        searchActions.updateQueryString(queryString),
    },
    dispatch
  );

const AdvancedSearchContainer = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Search)
);

export default AdvancedSearchContainer;
