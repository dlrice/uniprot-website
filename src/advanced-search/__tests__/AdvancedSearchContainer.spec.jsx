import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import createEmptyClause from '../utils/clause';
import { AdvancedSearch } from '../AdvancedSearchContainer';

configure({ adapter: new Adapter() });

let wrapper;
let props;

describe('AdvancedSearch component', () => {
  beforeEach(() => {
    props = {
      dispatchAddClause: jest.fn(),
      dispatchSubmitQuery: jest.fn(),
      dispatchFetchSearchTerms: jest.fn(),
      dispatchfetchEvidencesIfNeeded: jest.fn(),
      clauses: [...Array(4)].map(() => createEmptyClause()),
      namespace: 'UniProtKB',
      searchTerms: {
        data: [],
      },
      evidences: {
        go: {
          data: [],
          isFetching: false,
        },
        annotation: {
          data: [],
          isFetching: false,
        },
      },
    };
    wrapper = shallow(<AdvancedSearch {...props} />);
  });

  test('should call to get search terms', () => {
    expect(props.dispatchFetchSearchTerms).toHaveBeenCalled();
  });

  test('should call to get evidences', () => {
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('go');
    expect(props.dispatchfetchEvidencesIfNeeded).toHaveBeenCalledWith('annotation');
  });

  test('should submit a query', () => {
    wrapper.find('#submit-query').simulate('click');
    expect(props.dispatchSubmitQuery).toHaveBeenCalled();
  });

  test('should render', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should add field rows', () => {
    wrapper.find('#add-field').simulate('click');
    expect(props.dispatchAddClause).toHaveBeenCalled();
  });
});
