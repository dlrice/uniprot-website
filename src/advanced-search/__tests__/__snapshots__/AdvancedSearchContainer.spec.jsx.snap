// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch shallow components should render 1`] = `
"<div className=\\"advanced-search\\">
  <div>
    <label htmlFor=\\"namespace-select\\">
      Searching in
      <select id=\\"namespace-select\\">
        <option>
          UniProtKB
        </option>
      </select>
    </label>
  </div>
  <ClauseList dispatchAddClause={[Function: mockConstructor]} dispatchSubmitQuery={[Function: mockConstructor]} dispatchFetchSearchTerms={[Function: mockConstructor]} dispatchfetchEvidencesIfNeeded={[Function: mockConstructor]} dispatchCopyQueryClausesToSearch={[Function: mockConstructor]} history={{...}} clauses={{...}} namespace=\\"UniProtKB\\" searchTerms={{...}} evidences={{...}} />
  <hr />
  <div>
    <button type=\\"button\\" id=\\"add-field\\" className=\\"button\\" onClick={[Function: mockConstructor]}>
      Add Field
    </button>
    <button type=\\"button\\" id=\\"submit-query\\" className=\\"button\\" onClick={[Function: bound handleSubmitClick]}>
      Search
    </button>
  </div>
</div>"
`;
